atc/engine/task_delegate.go

```diff
@@ -35,12 +52,212 @@ type taskDelegate struct {
 	build       db.Build
 	eventOrigin event.Origin
 	clock       clock.Clock
+
+	dbWorkerFactory db.WorkerFactory
+	lockFactory     lock.LockFactory
 }
 
 func (d *taskDelegate) SetTaskConfig(config atc.TaskConfig) {
 	d.config = config
 }
 
+func (d *taskDelegate) SelectWorker(
+	ctx context.Context,
+	pool worker.Pool,
+	owner db.ContainerOwner,
+	containerSpec worker.ContainerSpec,
+	workerSpec worker.WorkerSpec,
+	strategy worker.ContainerPlacementStrategy,
+	workerPollingInterval, workerStatusPublishInterval time.Duration,
+) (worker.Client, error) {
+	logger := lagerctx.FromContext(ctx)
+
+	started := time.Now()
+	workerPollingTicker := time.NewTicker(workerPollingInterval)
+	defer workerPollingTicker.Stop()
+	workerStatusPublishTicker := time.NewTicker(workerStatusPublishInterval)
+	defer workerStatusPublishTicker.Stop()
+
+	tasksWaitingLabels := metric.TasksWaitingLabels{
+		TeamId:     strconv.Itoa(workerSpec.TeamID),
+		WorkerTags: strings.Join(workerSpec.Tags, "_"),
+		Platform:   workerSpec.Platform,
+	}
+
+	trySelectWorker := func() (worker.Client, error) {
+		var (
+			activeTasksLock lock.Lock
+			lockAcquired    bool
+			err             error
+		)
+		if strategy.ModifiesActiveTasks() {
+			for {
+				activeTasksLock, lockAcquired, err = d.lockFactory.Acquire(logger, lock.NewActiveTasksLockID())
+				if err != nil {
+					return nil, err
+				}
+
+				if lockAcquired {
+					defer activeTasksLock.Release()
+					break
+				}
+				// retry after a delay
+				time.Sleep(time.Second)
+			}
+		}
+
+		chosenWorker, err := pool.SelectWorker(
+			ctx,
+			owner,
+			containerSpec,
+			workerSpec,
+			strategy,
+		)
+		if err != nil {
+			// only the limit-active-tasks placement strategy waits for a
+			// worker to become available. All others should error out for now
+			allWorkersFullError := worker.NoWorkerFitContainerPlacementStrategyError{Strategy: "limit-active-tasks"}
+			if !errors.Is(err, allWorkersFullError) {
```

Comment:
@taylorsilva this logic is changed from #6359 - it seemed like it ignored *all* errors when selecting a worker, which likely isn't desirable, since it'd mask e.g database errors. Let me know if this should be changed back

Reply:
I think the difference is that previously, if the `limit-active-tasks` strategy was active, if the worker pool returned *any* error, we would ignore it. That means ignoring `NoWorkerFitContainerPlacementStrategyError`, but it also means ignoring legitimate errors (e.g. failed to lookup workers in the DB due to a network issue).

Side note: I mentioned this in standup briefly, but I think this inadvertently solved the `limit-active-tasks` issue we were seeing on CI. I *think* the issue was due to the active tasks lock not being released when selecting a worker failed:

https://github.com/concourse/concourse/blob/c371b460ba0b3842815d0f7bdc1bb1a373ea9294/atc/worker/client.go#L616-L625

If `strategy.ModifiesActiveTasks()`, then we return the error, but we never `defer`red the release of the lock, so we missed a branch.

However, ignoring *any* error from `pool.FindOrChooseWorkerForContainer` when using the `limit-active-tasks` strategy means that we never return if we acquired the lock (since we only acquire the lock for `limit-active-tasks`).

I refactored this code a bit so we're able to do a `defer lock.Release()` right next to the acquisition, which is a bit more sane than trying to do it on every branch